ygot/pathstrings.go

```diff
@@ -0,0 +1,268 @@
+			}
+		}
+		p = append(p, elem)
+	}
+	return filepath.Join(p...), nil
+}
+
+// PathType is used to indicate a gNMI path type.
+type PathType int64
+
+const (
+	// StructuredPath represents a Path using the structured 'PathElem' message in
+	// the 'elem' field of the gNMI Path message.
+	StructuredPath PathType = iota
+	// StringSlicePath represents a Path using the 'Element' repeated string field
+	// of the gNMI path message.
+	StringSlicePath
+)
+
+// StringToPath takes an input string representing a path in gNMI, and converts
+// it to a gNMI Path message, populated with the specified path encodings.
+func StringToPath(path string, pathTypes ...PathType) (*gnmipb.Path, error) {
+	var errs util.Errors
+	if len(pathTypes) == 0 {
+		return nil, util.AppendErr(errs, errors.New("no path types specified"))
+	}
+
+	pmsg := &gnmipb.Path{}
+	for _, p := range pathTypes {
+		switch p {
+		case StructuredPath:
+			gp, err := StringToStructuredPath(path)
+			if err != nil {
+				errs = util.AppendErr(errs, fmt.Errorf("error building structured path: %v", err))
+				continue
+			}
+			pmsg.Elem = gp.Elem
+		case StringSlicePath:
+			gp, err := StringToStringSlicePath(path)
+			if err != nil {
+				errs = util.AppendErr(errs, fmt.Errorf("error building string slice path: %v", err))
+				continue
+			}
+			pmsg.Element = gp.Element
+		}
+	}
+
+	if errs != nil {
+		return nil, errs
+	}
+
+	return pmsg, nil
+}
+
+// StringToStringSlicePath takes a string representing a path, and converts it into a
+// gnmi.Path. For example, if the Path "/a/b[c=d]/e" is input, it is converted
+// to a gnmi.Path{Element: []string{"a", "b[c=d]", "e"}} which is returned.
+func StringToStringSlicePath(path string) (*gnmipb.Path, error) {
+	parts := pathStringToElements(path)
+	for _, p := range parts {
+		// Run through extractKV to ensure that the path is valid.
+		if _, _, err := extractKV(p); err != nil {
+			return nil, fmt.Errorf("error parsing path %s: %v", path, err)
+		}
+	}
+
+	return &gnmipb.Path{
+		Element: parts,
+	}, nil
+}
+
+// StringToStructuredPath takes a string representing a path, and converts it to
+// a gnmi.Path, using the PathElem element message that is defined in gNMI 0.4.0.
+func StringToStructuredPath(path string) (*gnmipb.Path, error) {
+	parts := pathStringToElements(path)
+
+	gpath := &gnmipb.Path{}
+	for _, p := range parts {
+		name, kv, err := extractKV(p)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing path %s: %v", path, err)
+		}
+		gpath.Elem = append(gpath.Elem, &gnmipb.PathElem{
+			Name: name,
+			Key:  kv,
+		})
+	}
+	return gpath, nil
+}
+
+// pathStringToElements splits the string s, which represents a gNMI string
+// path into its constituent elements. It does not parse keys, which are left
+// unchanged within the path - but removes escape characters from element
+// names. The path returned omits any leading empty elements when splitting
+// on the / character.
+func pathStringToElements(s string) []string {
```

Comment:
I think just stripping \ could be dangerous e.g. what to do with \123.

Reply:
I'm not sure that this is a problem if you have `s := "\123gef"` then `range s` will resolve the control characters like this. See: https://play.golang.org/p/2KGZSMpRzH